### PR TITLE
Added sortkeys option to yaml snapshot generation

### PR DESCRIPTION
**Pull request by keesvanbemmel**
This new `--sortyaml` will sort the snapshot yaml alphabetically. ~~The downside is that the yaml itself is less readable, as all keys are sorted (e.g. the directus version isn't at the top anymore)~~.(Fixed downside by keeping the original top level order in place.)

The upside is that for really large snapshots (where sometimes the order of fields changes based on fields being added etc) it is now easy to compare diffs.

### CODE REVIEW

**Comment by azrikahar** [comment]
Understandably this circles back to the discussion over at #9198, but would you be able to provide a clear before/after example for comparison here? I'm sure it'll be helpful for other reviewers as well 👍

Additionally, upon further inspection at the options for `dump` (aliased to `toYaml`) of js-yaml [here](https://github.com/nodeca/js-yaml#dump-object---options-), seems like it does offer the use a function as well.

Relevant source code of the sortKeys: 

```js
  if (state.sortKeys === true) {
    // Default sorting
    objectKeyList.sort();
  } else if (typeof state.sortKeys === 'function') {
    // Custom sort function
    objectKeyList.sort(state.sortKeys);
  } 
```
  
_Link to above source code: https://github.com/nodeca/js-yaml/blob/49baadd52af887d2991e2c39a6639baa56d6c71b/lib/dumper.js#L684-L691_

I'm wondering perhaps we can opt to use a custom sort function that ensures `directus`, `version`, `collections`, `fields` etc retains their current order at the top most level, while the rest default back to the alphabetical sort. That way we'd be able to address the concerns back in #9198 from all angle.

**Comment by keesvanbemmel** [comment]
An example is hard, because in a plain vanilla setup it's hard to reproduce. And in our current environment I'm unable/unwilling to share our complete schema. But take this screenshot for example:
CleanShot 2022-01-07 at 11 52 07

Someone upgraded directus, merged the new snapshot to main
Someone added one field/collection to the schema based on the "old" schema in git.

The schema shows a lot of diffs because the order has changed. Mind you: **ALL** the diffs in the screenshot or also in that comparison the other way around. So there's nothing new in that screenshot. Everything there existed in the old and new schema.

I'm all for making a custom sort, making it even prettier. But for now having the unsorted way as a default, this PR will definitely help larger dev teams in the short run. If it were up to me I'd merge this and see/think/create a new custom sort at a later stage. Mind you, the default way is still the unsorted neat way.

I'll add the sortkeys to the json snapshot as well!

**Comment by azrikahar** [comment]
@keesvanbemmel thanks for the detailed explanation! IIRC I also have experienced this shifting of fields causing multiple unneeded diffs, and I do personally agree we should fix that for a better schema generation experience. 

However, I'm just wondering & I could be wrong here, since the default sort in js-yaml sorts by the object keys and not the values, wouldn't this shifting of fields still happen after we enabled the sort? 🤔

**Comment by keesvanbemmel** [comment]
Added a custom sort function. This is the before and after (small part of my schema):
BEFORE:
```
version: 1
directus: 9.4.2
collections:
  - collection: addresses
    meta:
      collection: addresses
      icon: null
      note: null
      display_template: null
      hidden: false
      singleton: false
      translations: null
      archive_field: status
      archive_app_filter: true
      archive_value: archived
      unarchive_value: draft
      sort_field: sort
      accountability: all
      color: null
      item_duplication_fields: null
      sort: null
      group: null
      collapse: open
    schema:
      name: addresses
      schema: public
      comment: null
  - collection: config
    meta:
      collection: config
      icon: null
      note: null
```

AFTER:
```
version: 1
directus: 9.4.2
collections:
  - collection: addresses
    meta:
      accountability: all
      archive_app_filter: true
      archive_field: status
      archive_value: archived
      collapse: open
      collection: addresses
      color: null
      display_template: null
      group: null
      hidden: false
      icon: null
      item_duplication_fields: null
      note: null
      singleton: false
      sort: null
      sort_field: sort
      translations: null
      unarchive_value: draft
    schema:
      comment: null
      name: addresses
      schema: public
  - collection: config
    meta:
      accountability: all
```

Hope this helps! It'll help me greatly 👯‍♀️

**Comment by keesvanbemmel** [comment]
> @keesvanbemmel thanks for the detailed explanation! IIRC I also have experienced this shifting of fields causing multiple unneeded diffs, and I do personally agree we should fix that for a better schema generation experience.
> 
> However, I'm just wondering & I could be wrong here, since the default sort in js-yaml sorts by the object keys and not the values, wouldn't this shifting of fields still happen after we enabled the sort? 🤔

@azrikahar you make a good point, I'm unsure, but I get the feeling it will be even far better now that the keys are sorted. There's no way now to test this unfortunately, reproducing is really hard.

File: api/src/cli/commands/schema/snapshot.ts [hunk 1]
```diff
@@ -7,9 +7,21 @@ import inquirer from 'inquirer';
 import { dump as toYaml } from 'js-yaml';
 import { flushCaches } from '../../../cache';
 
+/**
+ * Will sort the yaml keys by alphabetical order.
+ * If the key is a top level key, it will be sorted by its original order (defined in the topLevelKeys array).
+ */
+function sortYaml(a: any, b: any) {
+	const topLevelKeys = ['version', 'directus', 'collections', 'fields', 'relations'];
```
Comment (nickrum) [inline api/src/cli/commands/schema/snapshot.ts:15]: You could read those top level keys from the snapshot object instead of hardcoding them.

**Review by nickrum — CHANGES_REQUESTED** [review]
Still wondering at which point the keys are shuffled around. Though, as Javascript objects are unordered by design, it's probably a good idea in general to add this.
With the top level keys not being sorted, I think we could even remove the option and do this by default.
Also, we should probably do the same thing for json.

File: api/src/cli/commands/schema/snapshot.ts (hunk 1 above)
Comment (keesvanbemmel) [reply api/src/cli/commands/schema/snapshot.ts]: Just added commit to do that. Thanks!

**Comment by rijkvanzanten** [comment]
> Still wondering at which point the keys are shuffled around. Though, as Javascript objects are unordered by design, it's probably a good idea in general to add this.
> With the top level keys not being sorted, I think we could even remove the option and do this by default.
> Also, we should probably do the same thing for json.

Same here! The fields are read in the same way every time the operation is read, so there shouldn't be any deviation on what order the fields show up at any point 🤔

**Comment by keesvanbemmel** [comment]
Check this out after upgrading to 9.5.0 @rijkvanzanten 
CleanShot 2022-01-25 at 08 36 25

**Comment by rijkvanzanten** [comment]
> Check this out after upgrading to 9.5.0 @rijkvanzanten
> CleanShot 2022-01-25 at 08 36 25

That checks! Most likely because of the big upgrades on the knex-schema-inspector side. There was one other person who ran into a weird shuffle issue, so i'm curious to see if that's caused by a similar upstream change, or JS's lack of guaranteed object order (ref #11240)

**Pull request closed without merging**